### PR TITLE
meta: add swarm-datastore

### PR DESCRIPTION
**Pull request by orenyodfat**
This pr include the swarm-datastore meta interface .
This pr does not include cwr,musicbraniz,ern or the cli updates to use swarm-datastore.

### CODE REVIEW

File: meta.go [hunk 1]
```diff
@@ -246,6 +267,63 @@ func (s *Store) key(cid *cid.Cid) datastore.Key {
 	return datastore.NewKey(cid.String())
 }
 
+// NewSwarmStore returns a new Swarm Store which uses an underlying datastore.
+func NewSwarmStore(serverURL string) (*SwarmStore, error) {
+	store, err := swarmdatastore.NewDatastore(serverURL)
```
Comment (lmars) [inline meta.go]: Why do we need both `SwarmStore` and `swarmdatastore.Datastore`? Can't we use the the Swarm datastore with the existing `Store` type?

File: meta_test.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/ethereum/go-ethereum/swarm/testutil"
```
Comment (lmars) [inline meta_test.go]: Does this need to be vendored?

File: swarm-datastore/datastore.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package swarmdatastore
```
Comment (lmars) [inline swarm-datastore/datastore.go]: Does this need to be its own pacakge?

File: swarm-datastore/datastore.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	return datastore, err
+}
+
+// Put stores the given value and return its hash
+func (ds *Datastore) Put(value []byte) (hash string, err error) {
+	return ds.client.UploadRaw(bytes.NewReader(value), int64(len(value)))
+}
+
+// Get returns the value for given key
+func (ds *Datastore) Get(key string) (value interface{}, err error) {
```
Comment (lmars) [inline swarm-datastore/datastore.go]: Why does this return `interface{}`?

File: meta.go (hunk 1 above)
Comment (orenyodfat) [reply meta.go]: Why do we need both SwarmStore and swarmdatastore.Datastore?
 I though to keep it similar as possible to the ipfs datastore interface ..though we might not need that and can omit swarmdatastore.Datastore and be only with SwarmStore.
SwarmStore provides a bit different interface from ipfs Store . what will be the nicer way to merge it to one datastore interface ?

Comment (lmars) [reply meta.go]: I'm a little confused as to what you mean.

So we currently have the "META store" which has `Get(cid)` and `Put(obj)` and that wraps an underlying datastore interface which has `Get(key)` and `Put(key, value)`.

You have introduced both a "Swarm store" at the same level as the "META store" and also a "Swarm store" at the level of the underlying interface, and I am unsure why we need both?

I was expecting there to just be a new "Swarm store" as an underlying interface which could be used with the existing "META store", is that not possible?

**Comment by orenyodfat** [comment]
@lmars thanks! code updated

File: datastore.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	"golang.org/x/sync/syncmap"
+
+	"github.com/ethereum/go-ethereum/swarm/api/client"
+	multihash "github.com/multiformats/go-multihash"
+)
+
+// Datastore represents storage for any key-value pair.
+type Datastore interface {
+	put(data []byte) (mhash multihash.Multihash, err error)
+	get(key string) (value []byte, err error)
```
Comment (lmars) [inline datastore.go]: We should document that `key` should be the hex encoding of the hash part of the multihash returned from `put` (that isn't immediately obvious)

File: datastore.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+const multihashSwarmCode = 0x30
+
+func init() {
+	multihash.Codes[multihashSwarmCode] = "swarm-hash-v1"
+}
+
+// SwarmDatastore struct
+type SwarmDatastore struct {
+	serverURL string
```
Comment (lmars) [inline datastore.go]: Does this need to be a field?

File: datastore.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	if err != nil {
+		return false
+	}
+
+	return !finfo.IsDir()
+}
+
+// NullDatastore stores nothing, but conforms to the API.
+// Useful to test with.
+type NullDatastore struct {
```
Comment (lmars) [inline datastore.go]: Do we need this?

File: meta.go [hunk 8]
```diff
@@ -161,12 +164,12 @@ func (o *Object) MarshalJSON() ([]byte, error) {
 // Graph is used to traverse an object graph using a store and starting from
 // a particular root object.
 type Graph struct {
-	store *Store
+	store interface{}
```
Comment (lmars) [inline meta.go]: Why is this `interface{}`?

File: datastore.go (hunk 7 above)
Comment (orenyodfat) [reply datastore.go]: It is used by xml_test.go.

**Pull request merged**